### PR TITLE
add mysql driver

### PR DESCRIPTION
**Pull request by andxu**

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/utility/AzureDependencyActionBase.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import org.jetbrains.idea.maven.project.MavenProject;
+import org.jetbrains.idea.maven.project.MavenProjectsManager;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+public abstract class AzureDependencyActionBase extends AzureAnAction {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/utility/AzureDependencyActionBase.java:53]: `AzureDependencyActionBase` -> `AzureMavenDependencyActionBase`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/actions/mysql/AddMySQLDependencyAction.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+        DefaultLoader.getIdeHelper().runInBackground(project,
+                                                     "Add MySQL dependency",
+                                                     false,
+                                                     true,
+                                                     "Add MySQL dependency", () -> {
+                ProgressIndicator progressIndicator = ProgressManager.getInstance().getProgressIndicator();
+                progressIndicator.setText("Syncing maven project " + project.getName());
+                if (!syncMavenDependency(projectsManager, project, mavenProject)) {
+                    return;
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/actions/mysql/AddMySQLDependencyAction.java:83]: do we need to prompt any messages here?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/actions/mysql/AddMySQLDependencyAction.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                try {
+                    final String evaluateEffectivePom = MavenUtils.evaluateEffectivePom(project, mavenProject);
+                    ProgressManager.checkCanceled();
+
+                    if (StringUtils.isEmpty(evaluateEffectivePom)) {
+                        PluginUtil.showErrorNotificationProject(project, "Error", "Failed to evaluate effective pom.");
+                        return;
+                    }
+                    progressIndicator.setText("Get latest versions ...");
+                    SpringCloudDependencyManager dependencyManager = new SpringCloudDependencyManager(evaluateEffectivePom);
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/actions/mysql/AddMySQLDependencyAction.java:95]: Do we share codes with `SpringCloudDependencyManager`? If so , shall we rename it to `MavenDependencyManager`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/actions/mysql/AddMySQLDependencyAction.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    protected String getServiceName(AnActionEvent event) {
+        return TelemetryConstants.MYSQL;
+    }
+
+    @Override
+    protected String getOperationName(AnActionEvent event) {
+        return TelemetryConstants.ADD_MY_SQL_DEPENDENCY;
+    }
+
+    private static boolean applyVersionChanges(SpringCloudDependencyManager dependencyManager,
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/actions/mysql/AddMySQLDependencyAction.java:161]: could we extract this method to `AzureDependencyActionBase`? Seems the logic could be used for other dependency action

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/actions/mysql/AddMySQLDependencyAction.java (hunk 2 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/actions/mysql/AddMySQLDependencyAction.java:83]: syncMavenDependency will show error when syncing error

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/actions/mysql/AddMySQLDependencyAction.java (hunk 3 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/actions/mysql/AddMySQLDependencyAction.java:95]: No, spring cloud have complex and special logic for handling spring boot versions.

File: PluginsAndFeatures/azure-toolkit-for-intellij/resources/META-INF/plugin.xml [hunk 5]
```diff
@@ -291,9 +291,12 @@
             text="SparkJobRun"/>
     <action id="SparkJobDebug" class="com.microsoft.azure.hdinsight.spark.run.action.SparkJobDebugAction"
             text="SparkJobDebug"/>
-    <action id="Actions.AddAzureDependencyAction" class="com.microsoft.azure.springcloud.dependency.action.AddAzureDependencyAction" text="Add Azure Spring Cloud dependency"  icon="/icons/azure-springcloud-app-dependency.png">
+    <action id="Actions.AddAzureDependencyAction" class="com.microsoft.azure.springcloud.dependency.action.AddSpringClouMavenDependencyAction" text="Add Azure Spring Cloud dependency" icon="/icons/azure-springcloud-app-dependency.png">
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/resources/META-INF/plugin.xml]: how about to rename the id property as well ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/actions/mysql/AddMySQLDependencyAction.java (hunk 4 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/actions/mysql/AddMySQLDependencyAction.java]: Done

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/utility/AzureDependencyActionBase.java (hunk 1 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/utility/AzureDependencyActionBase.java]: Done

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/actions/mysql/AddMySQLDependencyAction.java (hunk 3 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/actions/mysql/AddMySQLDependencyAction.java]: SpringCloudDependencyManager extends MavenDependencyManager

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/utility/AzureAnAction.java [hunk 6]
```diff
@@ -88,6 +88,28 @@ public void sendTelemetryOnAction(AnActionEvent anActionEvent, final String acti
...
+    protected String getServiceName(AnActionEvent event) {
+        return TelemetryConstants.ACTION;
+    }
+
+    protected String getOperationName(AnActionEvent event) {
+        try {
+            return event.getPresentation().getText().toLowerCase().trim().replace(" ", "-");
+        } catch (Exception ignore) {
+            return "";
+        }
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/utility/AzureAnAction.java:110]: replace "", "" with StringUtils ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/maven/MavenDependencyManager.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        doc = DocumentHelper.parseText(effectivePomXml);
+    }
+
+    public String getPluginConfiguration(String groupId, String artifactId, String configurationName) {
+        for (Node node : doc.selectNodes("//ns:project/ns:build/ns:plugins/ns:plugin")) {
+            String myGroupId = ((Element) node).elementTextTrim("groupId");
+            String myArtifactId = ((Element) node).elementTextTrim("artifactId");
+            if (StringUtils.equals(groupId, myGroupId) && StringUtils.equals(artifactId, myArtifactId)) {
+                Element configurationNode = ((Element) node).element("configuration");
+                return configurationNode == null ? null : configurationNode.elementTextTrim(configurationName);
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/maven/MavenDependencyManager.java:54]: Is there any constant variables for "groupId", "artifactId" and "configuration" ?
I guess there would be some other classes need to reference them.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/maven/MavenDependencyManager.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        return version1.compareTo(version2) >= 0;
+    }
+
+    protected static void collectDependencyVersionsFromNodes(List<Node> nodes, Map<String, DependencyArtifact> versionMap) {
+        for (Node node : nodes) {
+            String groupId = ((Element) node).elementTextTrim("groupId");
+            String artifactId = ((Element) node).elementTextTrim("artifactId");
+            DependencyArtifact artifact = new DependencyArtifact(groupId, artifactId, ((Element) node).elementTextTrim("version"));
+            versionMap.put(artifact.getKey(), artifact);
+        }
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/maven/MavenDependencyManager.java:88]: here is a case for previous comment.

**Review by Flanker32 — APPROVED** [review]

**Review by shenqianjin — APPROVED** [review]

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/utility/AzureAnAction.java (hunk 6 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/utility/AzureAnAction.java:110]: it is an legacy code, I just move the method sequence

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/maven/MavenDependencyManager.java (hunk 7 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/maven/MavenDependencyManager.java:54]: resolved

**Pull request merged**